softlayer/resource_softlayer_subnet.go

```diff
@@ -0,0 +1,384 @@
+	"github.com/softlayer/softlayer-go/filter"
+	"github.com/softlayer/softlayer-go/helpers/product"
+	"github.com/softlayer/softlayer-go/services"
+	"github.com/softlayer/softlayer-go/session"
+	"github.com/softlayer/softlayer-go/sl"
+)
+
+const (
+	SubnetMask = "id,addressSpace,subnetType,version,ipAddressCount," +
+		"networkIdentifier,cidr,note,endPointIpAddress[ipAddress],networkVlan[id],totalIpAddresses"
+)
+
+var (
+	// Map subnet types to product package keyname in SoftLayer_Product_Item
+	subnetPackageTypeMap = map[string]string{
+		"Static":   "ADDITIONAL_SERVICES_STATIC_IP_ADDRESSES",
+		"Portable": "ADDITIONAL_SERVICES_PORTABLE_IP_ADDRESSES",
+	}
+)
+
+func resourceSoftLayerSubnet() *schema.Resource {
+	return &schema.Resource{
+		Create:   resourceSoftLayerSubnetCreate,
+		Read:     resourceSoftLayerSubnetRead,
+		Update:   resourceSoftLayerSubnetUpdate,
+		Delete:   resourceSoftLayerSubnetDelete,
+		Exists:   resourceSoftLayerSubnetExists,
+		Importer: &schema.ResourceImporter{},
+
+		Schema: map[string]*schema.Schema{
+			"id": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			"network": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errs []error) {
+					network := v.(string)
+					if network != "PRIVATE" && network != "PUBLIC" {
+						errs = append(errs, errors.New(
+							"network should be either 'PRIVATE' or 'PUBLIC'"))
+					}
+					return
+				},
+			},
+
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errs []error) {
+					typeStr := v.(string)
+					if typeStr != "Portable" && typeStr != "Static" {
+						errs = append(errs, errors.New(
+							"type should be either Portable or Static."))
+					}
+					return
+				},
+			},
+
+			// IP version 4 or IP version 6
+			"ip_version": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errs []error) {
+					ipVersion := v.(int)
+					if ipVersion != 4 && ipVersion != 6 {
+						errs = append(errs, errors.New(
+							"ip version should be either 4 or 6."))
+					}
+					return
+				},
+			},
+
+			"capacity": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+
+			// vlan_id should be configured when type is "Portable"
+			"vlan_id": {
+				Type:          schema.TypeInt,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"endpoint_ip"},
+			},
+
+			// endpoint_ip should be configured when type is "Static"
+			"endpoint_ip": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"vlan_id"},
+			},
+
+			// Provides IP address/netmask format (ex. 10.10.10.10/28)
+			"subnet": &schema.Schema{
```

Comment:
`subnet` is a computed attribute. So, validation is not necessary. In SoftLayer Portal, that column name is `subnet` and VLAN resource uses `subnet` as an attribute name of `IP address/cidr` format. I can change the attribute name to more user-friendly name. I think that `cidr` can be used as a name of this attribute. But, SoftLayer API uses this field name to represent `cidr block`(ex. /25). How about `ip_cidr` or `ip_address_cidr`?